src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+                        array[i - 1].Include(disposable);
+                        return;
+                    }
+                }
+            }
+        }
+
+        public class DisposeScope : IDisposable
+        {
+            private readonly DisposeScopeManager _disposeScopeManager;
+
+            public DisposeScope(DisposeScopeManager disposeScopeManager)
+            {
+                _disposeScopeManager = disposeScopeManager;
+            }
+
+            /// <summary>
+            /// The disposables that are scheduled for disposing.
+            /// </summary>
+            public HashSet<IDisposable> Disposables { get; } = new();
+
+            public long TotalAllocatedSize =>
+                Disposables.OfType<torch.Tensor>().Sum(x => x.NumberOfElements * x.ElementSize);
+
+            /// <summary>
+            /// Includes a disposable in the scope - for tensors this is done automatically once the scope has been
+            /// created. Use this method to add additional disposables that should be disposed, but you typically
+            /// don't need to call this method.
+            /// </summary>
+            /// <param name="disposable">The disposable to keep in the scope</param>
+            /// <returns></returns>
+            public T Include<T>(T disposable) where T : IDisposable
+            {
+                lock (this) {
+                    Disposables.Add(disposable);
+                }
+
+                return disposable;
+            }
+
+            /// <summary>
+            /// Excludes a set of tensors/disposables from the current dispose scope, and moves it up to the outer
+            /// dispose scope, if one exists. See overloaded methods. If you wish to exclude a tensor from all sccopes,
+            /// use ExcludeGlobally.
+            /// </summary>
+            public T Exclude<T>(T exclude) where T : IDisposable
+            {
+                Exclude(new IDisposable[] { exclude }, false);
+                return exclude;
+            }
+
+            public (T1 first, T2 second) Exclude<T1, T2>(T1 first, T2 second)
+                where T1 : IDisposable where T2 : IDisposable
+            {
+                Exclude(new IDisposable[] { first, second }, false);
+                return (first, second);
+            }
+
+            public (T1 first, T2 second, T3 third) Exclude<T1, T2, T3>(T1 first, T2 second, T3 third)
+                where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+            {
+                Exclude(new IDisposable[] { first, second, third }, false);
+                return (first, second, third);
+            }
+
+            public IDisposable[] Exclude(IDisposable[] exclude, bool excludeGlobally)
+            {
+                lock (this) {
+                    foreach (var disposable in exclude) {
+                        if (Disposables.Contains(disposable)) {
+                            Disposables.Remove(disposable);
```

Comment:
`HashSet.Remove` returns `true`/`false` depending if the item was found and removed.
Therefore you don't need tot call `Contains` first.

Reply:
Fixed, where I need to know if it actually existed, I use the return value of "Remove" as to not ask twice.